File: app/src/main/java/com/keylesspalace/tusky/util/HttpHeaderLink.java
/* Written in 2017 by Andrew Dawson
 *
 * To the extent possible under law, the author(s) have dedicated all copyright and related and
 * neighboring rights to this software to the public domain worldwide. This software is distributed
 * without any warranty.
 *
 * You should have received a copy of the CC0 Public Domain Dedication along with this software.
 * If not, see <http://creativecommons.org/publicdomain/zero/1.0/>. */

package com.keylesspalace.tusky.util;

import android.net.Uri;
import android.support.annotation.Nullable;

import java.util.ArrayList;
import java.util.List;

/**
 * Represents one link and its parameters from the link header of an HTTP message.
 *
 * @see <a href="https://tools.ietf.org/html/rfc5988">RFC5988</a>
 */
public class HttpHeaderLink {
    private static class Parameter {
        public String name;
        public String value;
    }

    private List<Parameter> parameters;
    public Uri uri;

    private HttpHeaderLink(String uri) {
        this.uri = Uri.parse(uri);
        this.parameters = new ArrayList<>();
    }

    private static int findAny(String s, int fromIndex, char[] set) {
        for (int i = fromIndex; i < s.length(); i++) {
            char c = s.charAt(i);
            for (char member : set) {
                if (c == member) {
                    return i;
                }
            }
        }
        return -1;
    }

    private static int findEndOfQuotedString(String line, int start) {
        for (int i = start; i < line.length(); i++) {
            char c = line.charAt(i);
            if (c == '\\') {
                i += 1;
            } else if (c == '"') {
                return i;
            }
        }
        return -1;
    }

    private static class ValueResult {
        String value;
        int end;

        ValueResult() {
            end = -1;
        }

        void setValue(String value) {
            value = value.trim();
            if (!value.isEmpty()) {
                this.value = value;
            }
        }
    }

    private static ValueResult parseValue(String line, int start) {
        ValueResult result = new ValueResult();
        int foundIndex = findAny(line, start, new char[] {';', ',', '"'});
        if (foundIndex == -1) {
            result.setValue(line.substring(start));
            return result;
        }
        char c = line.charAt(foundIndex);
        if (c == ';' || c == ',') {
            result.end = foundIndex;
            result.setValue(line.substring(start, foundIndex));
            return result;
        } else {
            int quoteEnd = findEndOfQuotedString(line, foundIndex + 1);
            if (quoteEnd == -1) {
                quoteEnd = line.length();
            }
            result.end = quoteEnd;
            result.setValue(line.substring(foundIndex + 1, quoteEnd));
            return result;
        }
    }

    private static int parseParameters(String line, int start, HttpHeaderLink link) {
        for (int i = start; i < line.length(); i++) {
            int foundIndex = findAny(line, i, new char[] {'=', ','});
            if (foundIndex == -1) {
                return -1;
            } else if (line.charAt(foundIndex) == ',') {
                return foundIndex;
            }
            Parameter parameter = new Parameter();
            parameter.name = line.substring(line.indexOf(';', i) + 1, foundIndex).trim();
            link.parameters.add(parameter);
            ValueResult result = parseValue(line, foundIndex);
            parameter.value = result.value;
            if (result.end == -1) {
                return -1;
            } else {
                i = result.end;
            }
        }
        return -1;
    }

    /**
     * @param line the entire link header, not including the initial "Link:"
     * @return all links found in the header
     */
    public static List<HttpHeaderLink> parse(@Nullable String line) {
        List<HttpHeaderLink> linkList = new ArrayList<>();
        if (line != null) {
            for (int i = 0; i < line.length(); i++) {
                int uriEnd = line.indexOf('>', i);
                String uri = line.substring(line.indexOf('<', i) + 1, uriEnd);
                HttpHeaderLink link = new HttpHeaderLink(uri);
                linkList.add(link);
                int parseEnd = parseParameters(line, uriEnd, link);
                if (parseEnd == -1) {
                    break;
                } else {
                    i = parseEnd;
                }
            }
        }
        return linkList;
    }

    /**
     * @param links intended to be those returned by parse()
     * @param relationType of the parameter "rel", commonly "next" or "prev"
     * @return the link matching the given relation type
     */
    @Nullable
    public static HttpHeaderLink findByRelationType(List<HttpHeaderLink> links,
            String relationType) {
        for (HttpHeaderLink link : links) {
            for (Parameter parameter : link.parameters) {
                if (parameter.name.equals("rel") && parameter.value.equals(relationType)) {
                    return link;
                }
            }
        }
        return null;
    }
}

File: app/src/main/java/com/keylesspalace/tusky/util/HttpHeaderLink.kt
/* Copyright 2022 Tusky Contributors
 *
 * This file is a part of Tusky.
 *
 * This program is free software; you can redistribute it and/or modify it under the terms of the
 * GNU General Public License as published by the Free Software Foundation; either version 3 of the
 * License, or (at your option) any later version.
 *
 * Tusky is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
 * Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with Tusky; if not,
 * see <http://www.gnu.org/licenses>.
 */
package com.keylesspalace.tusky.util

import android.net.Uri
import androidx.annotation.VisibleForTesting
import androidx.core.net.toUri

/**
 * Represents one link and its parameters from the link header of an HTTP message.
 *
 * @see [RFC5988](https://tools.ietf.org/html/rfc5988)
 */
class HttpHeaderLink @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE) constructor(
    uri: String
) {
    data class Parameter(val name: String, val value: String?)

    private val parameters: MutableList<Parameter> = ArrayList()

    val uri: Uri = uri.toUri()

    private data class ValueResult(val value: String, val end: Int = -1)

    companion object {
        private fun findEndOfQuotedString(line: String, start: Int): Int {
            var i = start
            while (i < line.length) {
                val c = line[i]
                if (c == '\\') {
                    i += 1
                } else if (c == '"') {
                    return i
                }
                i++
            }
            return -1
        }

        private fun parseValue(line: String, start: Int): ValueResult {
            val foundIndex = line.indexOfAny(charArrayOf(';', ',', '"'), start, false)
            if (foundIndex == -1) {
                return ValueResult(line.substring(start).trim())
            }
            val c = line[foundIndex]
            return if (c == ';' || c == ',') {
                ValueResult(line.substring(start, foundIndex).trim(), foundIndex)
            } else {
                var quoteEnd = findEndOfQuotedString(line, foundIndex + 1)
                if (quoteEnd == -1) {
                    quoteEnd = line.length
                }
                ValueResult(line.substring(foundIndex + 1, quoteEnd).trim(), quoteEnd)
            }
        }

        private fun parseParameters(line: String, start: Int, link: HttpHeaderLink): Int {
            var i = start
            while (i < line.length) {
                val foundIndex = line.indexOfAny(charArrayOf('=', ','), i, false)
                if (foundIndex == -1) {
                    return -1
                } else if (line[foundIndex] == ',') {
                    return foundIndex
                }
                val name = line.substring(line.indexOf(';', i) + 1, foundIndex).trim()
                val result = parseValue(line, foundIndex)
                val value = result.value
                val parameter = Parameter(name, value)
                link.parameters.add(parameter)
                i = if (result.end == -1) {
                    return -1
                } else {
                    result.end
                }
            }
            return -1
        }

        /**
         * @param line the entire link header, not including the initial "Link:"
         * @return all links found in the header
         */
        fun parse(line: String?): List<HttpHeaderLink> {
            val links: MutableList<HttpHeaderLink> = mutableListOf()
            line ?: return links

            var i = 0
            while (i < line.length) {
                val uriEnd = line.indexOf('>', i)
                val uri = line.substring(line.indexOf('<', i) + 1, uriEnd)
                val link = HttpHeaderLink(uri)
                links.add(link)
                val parseEnd = parseParameters(line, uriEnd, link)
                i = if (parseEnd == -1) {
                    break
                } else {
                    parseEnd
                }
                i++
            }

            return links
        }

        /**
         * @param links intended to be those returned by parse()
         * @param relationType of the parameter "rel", commonly "next" or "prev"
         * @return the link matching the given relation type
         */
        fun findByRelationType(
            links: List<HttpHeaderLink>,
            relationType: String
        ): HttpHeaderLink? {
            return links.find { link ->
                link.parameters.any { parameter ->
                    parameter.name == "rel" && parameter.value == relationType
                }
            }
        }
    }
}
